docs/src/integrations/exchange.md

```diff
@@ -0,0 +1,385 @@
+a transfer transaction, and submit it using the `sendAndConfirmTransaction`
+method.
+
+### Asynchronous
+
+For greater flexibility, you can submit withdrawal transfers asynchronously. In
+these cases, it is your responsibility to verify that the transaction succeeded
+and was finalized by the cluster.
+
+**Note:** Each transaction contains a [recent
+blockhash](../transaction.md#blockhash-format) to indicate its liveness. It is
+**critical** to wait until this blockhash expires before retrying a withdrawal
+transfer that does not appear to have been confirmed or finalized by the
+cluster. Otherwise, you risk a double spend. See [blockhash expiration](#blockhash-expiration)
+below.
+
+In the command-line tool, pass the `--no-wait` argument to send a transfer
+asynchronously:
+
+```bash
+solana transfer <USER_ADDRESS> <AMOUNT> --no-wait --keypair <KEYPAIR> --url http://localhost:8899
+```
+
+You can also build, sign, and serialize the transaction manually, and fire it off to
+the cluster using the JSON-RPC [`sendTransaction` endpoint](../apps/jsonrpc-api.md#sendtransaction).
+
+#### Transaction Confirmations & Finality
+
+Get the status of a batch of transactions using the
+[`getSignatureStatuses` JSON-RPC endpoint](../apps/jsonrpc-api.md#getsignaturestatuses).
+The `confirmations` field reports how many
+[confirmed blocks](../terminology.md#confirmed-block) have elapsed since the
+transaction was processed. If `confirmations: null`, it is [finalized](../terminology.md#finality).
+
+```bash
+curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "method":"getSignatureStatuses", "params":[["5VERv8NMvzbJMEkV8xnrLkEaWRtSz9CosKDYjCJjBRnbJLgp8uirBgmQpjKhoR4tjF3ZpRzrFmBV6UjKdiSZkQUW", "5j7s6NiJS3JAkvgkoc18WVAsiSaci2pxB2A6ueCJP4tprA2TFg9wSyTLeYouxPBJEMzJinENTkpA52YStRW5Dia7"]]}' http://localhost:8899
+
+{
+  "jsonrpc": "2.0",
+  "result": {
+    "context": {
+      "slot": 82
+    },
+    "value": [
+      {
+        "slot": 72,
+        "confirmations": 10,
+        "err": null,
+        "status": {
+          "Ok": null
+        }
+      },
+      {
+        "slot": 48,
+        "confirmations": null,
+        "err": null,
+        "status": {
+          "Ok": null
+        }
+      }
+    ]
+  },
+  "id": 1
+}
+```
+
+#### Blockhash Expiration
+
+To check whether a recent blockhash is still valid, send a
+[`getFeeCalculatorForBlockhash`](../apps/jsonrpc-api.md#getfeecalculatorforblockhash)
+request with the blockhash as a parameter. If the response value is null, the
+blockhash is expired, and the withdrawal transaction should never succeed.
+
+You can also use the
+[`getBlockhashLifespan` endpoint](../apps/jsonrpc-api.md#getblockhashlifespan)
+to determine how many slots a blockhash is valid for in the current epoch. When
+you request a [recent blockhash](../apps/jsonrpc-api.md#getrecentblockhash), the
+response includes the slot context. The blockhash is valid until `blockhashSlot +
```

Comment:
What precisely is the slot in the slot context? I'm concerned there's a difference between that slot and the slot the blockhash was pushed into the queue.

Reply:
Also, I see a race condition here.  If `getBlockhashLifespan` is called after `getRecentBlockhash`, the lifespan may have changed, and the client may resign and send a new transaction and then find both to get accepted.  I think the lifespan (or "valid through slot height") data needs to be returned alongside the blockhash.

Reply:
It is the slot of the bank that was queried to get the recent blockhash. In fact there is a small difference between that slot and the slot the blockhash was pushed into the queue, because of this code here: https://github.com/solana-labs/solana/blob/486168b79641bac0fa22e929ce5fdba3fce3f231/runtime/src/bank.rs#L995

But if we go with your suggestion (https://github.com/solana-labs/solana/pull/10162#issuecomment-632328404), and calculate valid_through_slot_height precisely based on the hash's position in the queue, this fuzziness goes away.

Reply:
✔️

Reply:
Thanks!